ch_7_Finetune/7_1_Finetune_Llama2-7B.md

```diff
@@ -0,0 +1,321 @@
+
+## 7.1.1 Enable BigDL-LLM on Intel GPUs
+
+
+### 7.1.1.1 Install BigDL-LLM on Intel GPUs
+
+After following the steps in [Readme](./README.md#70-environment-setup)  to set up the environment, you can install BigDL-LLM in terminal with the command below:
+```bash
+pip install bigdl-llm[xpu] -f https://developer.intel.com/ipex-whl-stable-xpu
+```
+
+> **Note**
+> The above command will install `intel_extension_for_pytorch==2.0.110+xpu` as default
+
+### 7.1.1.2 Import `intel_extension_for_pytorch`
+
+After installation, let's move to the **Python scripts** of this tutorial. First of all you need to import `intel_extension_for_pytorch` first for BigDL-LLM optimizations:
+
+```python
+import intel_extension_for_pytorch as ipex
+```
+
+## 7.1.2 QLoRA Finetuning
+
+### 7.1.2.1 Load Model in Low Precision
+
+
+A popular open-source LLM [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-hf) is chosen to illustrate the process of model finetuning.
+
+> **Note**
+>
+> You can specify the argument `pretrained_model_name_or_path` with both Huggingface repo id or local model path.
+> If you have already downloaded the Llama 2 (7B) model, you could specify `pretrained_model_name_or_path` to the local model path.
+
+With BigDL-LLM optimization, you can load the model with `bigdl.llm.transformers.AutoModelForCausalLM` instead of `transformers.AutoModelForCausalLM` to conduct implicit quantization.
+
+For Intel GPUs, once you have the model in low precision, **set it to `to('xpu')`**.
+
+```python
+model = AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path = "meta-llama/Llama-2-7b-hf",
+                                            load_in_low_bit="nf4",
+                                            optimize_model=False,
+                                            torch_dtype=torch.float16,
+                                            modules_to_not_convert=["lm_head"])
+model = model.to('xpu')
+
+```
+
+> **Note**
+>
+> We specify load_in_low_bit="nf4" here to apply 4-bit NormalFloat optimization. According to the [QLoRA paper](https://arxiv.org/pdf/2305.14314.pdf), using "nf4" could yield better model quality than "int4".
+
+### 7.1.2.2 Prepare Model for Training
+Then we apply `prepare_model_for_kbit_training` from `bigdl.llm.transformers.qlora` to preprocess the model for training. 
+
+```python
+from bigdl.llm.transformers.qlora import prepare_model_for_kbit_training
+model.gradient_checkpointing_enable()
+model = prepare_model_for_kbit_training(model)
+```
+
+Next, we can obtain a PEFT model from the optimized model and a configuration object containing the parameters as follows:  
+
+```python
+from bigdl.llm.transformers.qlora import get_peft_model
+from peft import LoraConfig
+
+config = LoraConfig(r=8, # defines how many parameters will be trained
```

Comment:
Please  update comments of these parameters, referring to https://huggingface.co/docs/peft/conceptual_guides/lora#common-lora-parameters-in-peft